### PR TITLE
CRM-17323 - Add sid token to registration/membership links

### PR DESCRIPTION
**Pull request by colemanw**
https://issues.civicrm.org/jira/browse/CRM-17323

### CODE REVIEW

File: CRM/Upgrade/Incremental/sql/4.6.9.mysql.tpl [hunk 1]
```diff
@@ -13,3 +13,44 @@ ADD COLUMN `created_id` int(10) unsigned DEFAULT NULL COMMENT 'FK to contact tab
...
+( {$domainID}, 'http://civicrm.org/documentation?src=iam',      '{ts escape="sql" skip="true"}Documentation{/ts}',      'Documentation',      NULL, 'AND', @adminHelplastID, '1', NULL, 2 ),
+( {$domainID}, 'http://civicrm.org/ask-a-question?src=iam',     '{ts escape="sql" skip="true"}Ask a question{/ts}',     'Ask a question',     NULL, 'AND', @adminHelplastID, '1', NULL, 3 ),
+( {$domainID}, 'http://civicrm.org/experts?src=iam',            '{ts escape="sql" skip="true"}Get expert help{/ts}',    'Get expert help',    NULL, 'AND', @adminHelplastID, '1', NULL, 4 ),
+( {$domainID}, 'http://civicrm.org/about?src=iam',              '{ts escape="sql" skip="true"}About CiviCRM{/ts}',      'About CiviCRM',      NULL, 'AND', @adminHelplastID, '1', 1, 5 ),
+( {$domainID}, 'http://civicrm.org/register-your-site?src=iam&sid={ldelim}sid{rdelim}', '{ts escape="sql" skip="true"}Register your site{/ts}', 'Register your site', NULL, 'AND', @adminHelplastID, '1', NULL, 6 ),
+( {$domainID}, 'http://civicrm.org/become-a-member?src=iam&sid={ldelim}sid{rdelim}',      '{ts escape="sql" skip="true"}Join CiviCRM{/ts}',       'Join CiviCRM',       NULL, 'AND', @adminHelplastID, '1', NULL, 7 );
```
Comment (totten) [inline CRM/Upgrade/Incremental/sql/4.6.9.mysql.tpl]: If someone upgraded 4.6.0=>4.6.6 (as previously tagged/published) then later 4.6.6=>4.6.9, wouldn't they get two records for each link (one from each upgrade)?

Comment (colemanw) [inline CRM/Upgrade/Incremental/sql/4.6.9.mysql.tpl]: No, because of the DELETE statement.

**Comment by colemanw** [comment]
@totten this is ready for review/merge. Once merged I will need to do some followup to ensure it merges cleanly into 4.7.

**Comment by colemanw** [comment]
Note to self:  upgrade success.tpl needs updating for 4.7.

File: CRM/Upgrade/Incremental/sql/4.6.9.mysql.tpl (hunk 1 above)
Comment (totten) [inline CRM/Upgrade/Incremental/sql/4.6.9.mysql.tpl]: Ah, good point.

It probably doesn't matter too much... but I think that DELETE might remove anything that sysadmin has put under "Help" manually. Maybe tighten it like this?

```sql
DELETE FROM civicrm_navigation WHERE parent_id = @adminHelplastID AND url LIKE "http://civicrm.org%"
```

File: CRM/Utils/System.php [hunk 2]
```diff
@@ -1787,18 +1787,31 @@ public static function evalUrl($url) {
...
   }
 
+  /**
+   * Returns the unique identifier for this site, as used by community messages.
+   *
+   * SiteID will be generated if it is not already stored in the settings table.
+   *
+   * @return string
+   */
+  public static function getSiteID() {
```
Comment (totten) [inline CRM/Utils/System.php]: Like the approach.

File: CRM/Upgrade/Incremental/sql/4.6.9.mysql.tpl (hunk 1 above)
Comment (colemanw) [inline CRM/Upgrade/Incremental/sql/4.6.9.mysql.tpl]: That's a good idea.
For that matter, all these links really ought to be https.

File: CRM/Utils/System.php (hunk 2 above)
Comment (colemanw) [inline CRM/Utils/System.php]: Yea, `getItem` uses static caching so I didn't bother with any other caching in this fn.

**Pull request merged**